policy/store.go

```diff
@@ -0,0 +1,206 @@
+	// TODO ensure uniqueness of datacenter/external ID combination.
+	// TODO assume that external ID is taken from name if not specified.
+	// At least one must be specified.
+	json, err := json.Marshal(policyDoc)
+	policyDb := &PolicyDb{}
+	policyDb.Policy = string(json)
+	policyDb.ExternalID = policyDoc.ExternalID
+	db := policyStore.DbStore.Db
+	db.Create(policyDb)
+	err = common.GetDbErrors(db)
+	if err != nil {
+		return err
+	}
+	db.NewRecord(*policyDb)
+	err = common.GetDbErrors(db)
+	if err != nil {
+		return err
+	}
+	policyDoc.ID = policyDb.ID
+	log.Printf("addPolicy(): Stored %s with ID %d", policyDoc.Name, policyDb.ID)
+	return nil
+}
+
+func (policyStore *policyStore) listPolicies() ([]common.Policy, error) {
+	var policyDb []PolicyDb
+	var policies []common.Policy
+	db := policyStore.DbStore.Db.Find(&policyDb)
+	err := common.GetDbErrors(db)
+	if err != nil {
+		return policies, err
+	}
+	policies = make([]common.Policy, len(policyDb))
+	for i, p := range policyDb {
+		json.Unmarshal([]byte(p.Policy), &policies[i])
+		policies[i].ID = p.ID
+	}
+	return policies, err
+}
+
+func (policyStore *policyStore) lookupPolicy(externalID string, dcID uint64) (uint64, error) {
+	policyDbEntry := PolicyDb{}
+	log.Printf("Looking up policy with id = %s ", externalID)
+	db := policyStore.DbStore.Db.First(&policyDbEntry, "external_id = ?", externalID)
+	if db.RecordNotFound() {
+		return 0, common.NewError404("policy", externalID)
+	}
+	err := common.GetDbErrors(db)
+	// TODO return proper error (404) in case not found.
+	if err != nil {
+		return 0, err
+	}
+	return policyDbEntry.ID, nil
+}
+
+func (policyStore *policyStore) getPolicy(id uint64, markedDeleted bool) (common.Policy, error) {
+	policyDbEntry := PolicyDb{}
+	policyDoc := common.Policy{}
+	log.Printf("Looking up policy with id = %v (deleted: %v)", id, markedDeleted)
+	var err error
+	if markedDeleted {
+		db := policyStore.DbStore.Db.Unscoped().First(&policyDbEntry, "id = ?", id)
+		if db.RecordNotFound() {
+			return policyDoc, common.NewError404("policy", string(id))
+		}
+		err = common.GetDbErrors(db)
+	} else {
+		db := policyStore.DbStore.Db.First(&policyDbEntry, "id = ?", id)
+		if db.RecordNotFound() {
+			return policyDoc, common.NewError404("policy", string(id))
+		}
+		err = common.GetDbErrors(db)
+	}
+	// TODO return proper error (404) in case not found.
+	if err != nil {
+		return policyDoc, err
+	}
+	log.Printf("Found %v, unmarshaling %s", policyDbEntry, policyDbEntry.Policy)
+	err = json.Unmarshal([]byte(policyDbEntry.Policy), &policyDoc)
+	if err != nil {
+		return policyDoc, err
+	}
+	policyDoc.ID = policyDbEntry.ID
+	return policyDoc, err
+}
+
+// inactivatePolicy marks policy as inactive. This is done
+// upon receiving a DELETE request but before distributing
+// this request to agents.
+func (policyStore *policyStore) inactivatePolicy(id uint64) error {
+	policyDb := &PolicyDb{}
+	db := policyStore.DbStore.Db
+	db = db.Where("id = ?", id).Delete(policyDb)
```

Comment:
This doesn't look like you are deactivating it. Instead, it looks like you are deleting it.